File: DriverTemplates/VSIXInstaller/DriverTemplate(VB)/DeviceObservingConditions.vb
' All lines from line 1 to the device interface implementation region will be discarded by the project wizard when the template is used
' Required code must lie within the device implementation region
' The //ENDOFINSERTEDFILE tag must be the last but one line in this file

Imports ASCOM.DeviceInterface
Imports ASCOM
Imports ASCOM.Utilities

Class DeviceObservingConditions
    Implements IObservingConditions

    Private TL As New TraceLogger()

#Region "IObservingConditions Implementation"

	''' <summary>
	''' Gets and sets the time period over which observations wil be averaged
	''' </summary>
	Public Property AveragePeriod() As Double Implements IObservingConditions.AveragePeriod
		Get
			TL.LogMessage("AveragePeriod", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("AveragePeriod", False)
		End Get
		Set(value As Double)
			TL.LogMessage("AveragePeriod", "Set Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("AveragePeriod", True)
		End Set
	End Property

	''' <summary>
	''' Amount of sky obscured by cloud
	''' </summary>
	Public ReadOnly Property CloudCover() As Double Implements IObservingConditions.CloudCover
		Get
			TL.LogMessage("CloudCover", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("CloudCover", False)
		End Get
	End Property

	''' <summary>
	''' Atmospheric dew point at the observatory in deg C
	''' </summary>
	Public ReadOnly Property DewPoint() As Double Implements IObservingConditions.DewPoint
		Get
			TL.LogMessage("DewPoint", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("DewPoint", False)
		End Get
	End Property

	''' <summary>
	''' Atmospheric relative humidity at the observatory in percent
	''' </summary>
	Public ReadOnly Property Humidity() As Double Implements IObservingConditions.Humidity
		Get
			TL.LogMessage("Humidity", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("Humidity", False)
		End Get
	End Property

	''' <summary>
	''' Atmospheric pressure at the observatory in hectoPascals (mB)
	''' </summary>
	Public ReadOnly Property Pressure() As Double Implements IObservingConditions.Pressure
		Get
			TL.LogMessage("Pressure", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("Pressure", False)
		End Get
	End Property

	''' <summary>
	''' Rain rate at the observatory
	''' </summary>
	Public ReadOnly Property RainRate() As Double Implements IObservingConditions.RainRate
		Get
			TL.LogMessage("RainRate", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("RainRate", False)
		End Get
	End Property

	''' <summary>
	''' Sky brightness at the observatory
	''' </summary>
	Public ReadOnly Property SkyBrightness() As Double Implements IObservingConditions.SkyBrightness
		Get
			TL.LogMessage("SkyBrightness", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("SkyBrightness", False)
		End Get
	End Property

	''' <summary>
	''' Sky quality at the observatory
	''' </summary>
	Public ReadOnly Property SkyQuality() As Double Implements IObservingConditions.SkyQuality
		Get
			TL.LogMessage("SkyQuality", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("SkyQuality", False)
		End Get
	End Property

	''' <summary>
	''' Seeing at the observatory
	''' </summary>
	Public ReadOnly Property StarFWHM() As Double Implements IObservingConditions.StarFWHM
		Get
			TL.LogMessage("StarFWHM", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("StarFWHM", False)
		End Get
	End Property

	''' <summary>
	''' Sky temperature at the observatory in deg C
	''' </summary>
	Public ReadOnly Property SkyTemperature() As Double Implements IObservingConditions.SkyTemperature
		Get
			TL.LogMessage("SkyTemperature", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("SkyTemperature", False)
		End Get
	End Property

	''' <summary>
	''' Temperature at the observatory in deg C
	''' </summary>
	Public ReadOnly Property Temperature() As Double Implements IObservingConditions.Temperature
		Get
			TL.LogMessage("Temperature", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("Temperature", False)
		End Get
	End Property

	''' <summary>
	''' Wind direction at the observatory in degrees
	''' </summary>
	Public ReadOnly Property WindDirection() As Double Implements IObservingConditions.WindDirection
		Get
			TL.LogMessage("WindDirection", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("WindDirection", False)
		End Get
	End Property

	''' <summary>
	''' Peak 3 second wind gust at the observatory over the last 2 minutes in m/s
	''' </summary>
	Public ReadOnly Property WindGust() As Double Implements IObservingConditions.WindGust
		Get
			TL.LogMessage("WindGust", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("WindGust", False)
		End Get
	End Property

	''' <summary>
	''' Wind speed at the observatory in m/s
	''' </summary>
	Public ReadOnly Property WindSpeed() As Double Implements IObservingConditions.WindSpeed
		Get
			TL.LogMessage("WindSpeed", "Get Not implemented")
			Throw New ASCOM.PropertyNotImplementedException("WindSpeed", False)
		End Get
	End Property

    ''' <summary>
    ''' Provides the time since the sensor value was last updated
    ''' </summary>
    ''' <param name="propertyName">Name of the property whose time since last update Is required</param>
    ''' <returns>Time in seconds since the last sensor update for this property</returns>
    Public Function TimeSinceLastUpdate(PropertyName As String) As Double Implements IObservingConditions.TimeSinceLastUpdate
        Dim lastUpdateTime As Double = 0.0

		' Test for an empty property name, if found, return the time since the most recent update to any sensor
		If Not String.IsNullOrEmpty(PropertyName) Then
			Select Case PropertyName.Trim.ToLowerInvariant
				' Return the time for properties that are implemented, otherwise fall through to the MethodNotImplementedException
				Case "averageperiod"
				Case "cloudcover"
				Case "dewpoint"
				Case "humidity"
				Case "pressure"
				Case "rainrate"
				Case "skybrightness"
				Case "skyquality"
				Case "skytemperature"
				Case "starfwhm"
				Case "temperature"
				Case "winddirection"
				Case "windgust"
				Case "windspeed"
					' Throw an exception on the properties that are Not implemented
					TL.LogMessage("TimeSinceLastUpdate", PropertyName & " - not implemented")
					Throw New MethodNotImplementedException("TimeSinceLastUpdate(" + PropertyName + ")")
				Case Else
					TL.LogMessage("TimeSinceLastUpdate", PropertyName & " - unrecognised")
					Throw New ASCOM.InvalidValueException("TimeSinceLastUpdate(" + PropertyName + ")")
			End Select
		Else
			' Return the time since the most recent update to any sensor
			TL.LogMessage("TimeSinceLastUpdate", $"The time since the most recent sensor update is not implemented")
			Throw New MethodNotImplementedException("TimeSinceLastUpdate(" + PropertyName + ")")
		End If

		Return lastUpdateTime
    End Function

    ''' <summary>
    ''' Provides a description of the sensor providing the requested property
    ''' </summary>
    ''' <param name="propertyName">Name of the property whose sensor description is required</param>
    ''' <returns>The sensor description string</returns>
    Public Function SensorDescription(PropertyName As String) As String Implements IObservingConditions.SensorDescription
		Select Case PropertyName.Trim.ToLowerInvariant
			Case "averageperiod"
				Return "Average period in hours, immediate values are only available"
			Case "cloudcover"
			Case "dewpoint"
			Case "humidity"
			Case "pressure"
			Case "rainrate"
			Case "skybrightness"
			Case "skyquality"
			Case "skytemperature"
			Case "starfwhm"
			Case "temperature"
			Case "winddirection"
			Case "windgust"
			Case "windspeed"
				' Throw an exception on the properties that are Not implemented
				TL.LogMessage("SensorDescription", $"Property {PropertyName} is not implemented")
				Throw New MethodNotImplementedException($"SensorDescription - Property {PropertyName} is not implemented")
		End Select

		TL.LogMessage("SensorDescription", $"Invalid sensor name: {PropertyName}")
		Throw New InvalidValueException($"SensorDescription - Invalid property name: {PropertyName}")
	End Function

	''' <summary>
	''' Forces the driver to immediately query its attached hardware to refresh sensor
	''' values
	''' </summary>
	Public Sub Refresh() Implements IObservingConditions.Refresh
		TL.LogMessage("Refresh", "Not implemented")
		Throw New ASCOM.MethodNotImplementedException("Refresh")
	End Sub

#End Region

	'//ENDOFINSERTEDFILE
End Class

File: DriverTemplates/VSIXInstaller/DriverTemplate(CS)/DeviceObservingConditions.cs
// All lines from line 1 to the device interface implementation region will be discarded by the project wizard when the template is used
// Required code must lie within the device implementation region
// The //ENDOFINSERTEDFILE tag must be the last but one line in this file

using ASCOM;
using ASCOM.Utilities;
using System;
using System.Collections.Generic;

class DeviceObservingConditions
{
    //private static TraceLogger tl = new TraceLogger();

    private static void LogMessage(string identifier, string message, params object[] args)
    {
        //tl.LogMessage(identifier, string.Format(message, args));
    }

    #region IObservingConditions Implementation

    /// <summary>
    /// Gets and sets the time period over which observations will be averaged
    /// </summary>
    public double AveragePeriod
    {
        get
        {
            LogMessage("AveragePeriod", "get - 0");
            return 0;
        }
        set
        {
            LogMessage("AveragePeriod", "set - {0}", value);
            if (value != 0)
                throw new InvalidValueException("AveragePeriod", value.ToString(), "0 only");
        }
    }

    /// <summary>
    /// Amount of sky obscured by cloud
    /// </summary>
    public double CloudCover
    {
        get
        {
            LogMessage("CloudCover", "get - not implemented");
            throw new PropertyNotImplementedException("CloudCover", false);
        }
    }

    /// <summary>
    /// Atmospheric dew point at the observatory in deg C
    /// </summary>
    public double DewPoint
    {
        get
        {
            LogMessage("DewPoint", "get - not implemented");
            throw new PropertyNotImplementedException("DewPoint", false);
        }
    }

    /// <summary>
    /// Atmospheric relative humidity at the observatory in percent
    /// </summary>
    public double Humidity
    {
        get
        {
            LogMessage("Humidity", "get - not implemented");
            throw new PropertyNotImplementedException("Humidity", false);
        }
    }

    /// <summary>
    /// Atmospheric pressure at the observatory in hectoPascals (hPa)
    /// </summary>
    public double Pressure
    {
        get
        {
            LogMessage("Pressure", "get - not implemented");
            throw new PropertyNotImplementedException("Pressure", false);
        }
    }

	/// <summary>
	/// Rain rate at the observatory, in millimeters per hour
	/// </summary>
	public double RainRate
    {
        get
        {
            LogMessage("RainRate", "get - not implemented");
            throw new PropertyNotImplementedException("RainRate", false);
        }
    }

    /// <summary>
    /// Forces the driver to immediately query its attached hardware to refresh sensor
    /// values
    /// </summary>
    public void Refresh()
    {
        throw new MethodNotImplementedException();
    }

    /// <summary>
    /// Provides a description of the sensor providing the requested property
    /// </summary>
    /// <param name="propertyName">Name of the property whose sensor description is required</param>
    /// <returns>The sensor description string</returns>
    public string SensorDescription(string propertyName)
    {
        switch (propertyName.Trim().ToLowerInvariant())
        {
            case "averageperiod":
                return "Average period in hours, immediate values are only available";
            case "cloudcover":
            case "dewpoint":
            case "humidity":
            case "pressure":
            case "rainrate":
            case "skybrightness":
            case "skyquality":
            case "skytemperature":
            case "starfwhm":
            case "temperature":
            case "winddirection":
            case "windgust":
            case "windspeed":
                // Throw an exception on the properties that are not implemented
                LogMessage("SensorDescription", $"Property {propertyName} is not implemented");
                throw new MethodNotImplementedException($"SensorDescription - Property {propertyName} is not implemented");
            default:
                LogMessage("SensorDescription", $"Invalid sensor name: {propertyName}");
                throw new InvalidValueException($"SensorDescription - Invalid property name: {propertyName}");
        }
    }

    /// <summary>
    /// Sky brightness at the observatory, in Lux (lumens per square meter)
    /// </summary>
    public double SkyBrightness
    {
        get
        {
            LogMessage("SkyBrightness", "get - not implemented");
            throw new PropertyNotImplementedException("SkyBrightness", false);
        }
    }

	/// <summary>
	/// Sky quality at the observatory, in magnitudes per square arc-second
	/// </summary>
	public double SkyQuality
    {
        get
        {
            LogMessage("SkyQuality", "get - not implemented");
            throw new PropertyNotImplementedException("SkyQuality", false);
        }
    }

	/// <summary>
	/// Seeing at the observatory, measured as the average star full width half maximum (FWHM in arc secs) 
	/// within a star field
	/// </summary>
	public double StarFWHM
    {
        get
        {
            LogMessage("StarFWHM", "get - not implemented");
            throw new PropertyNotImplementedException("StarFWHM", false);
        }
    }

    /// <summary>
    /// Sky temperature at the observatory in deg C
    /// </summary>
    public double SkyTemperature
    {
        get
        {
            LogMessage("SkyTemperature", "get - not implemented");
            throw new PropertyNotImplementedException("SkyTemperature", false);
        }
    }

    /// <summary>
    /// Temperature at the observatory in deg C
    /// </summary>
    public double Temperature
    {
        get
        {
            LogMessage("Temperature", "get - not implemented");
            throw new PropertyNotImplementedException("Temperature", false);
        }
    }

    /// <summary>
    /// Provides the time since the sensor value was last updated
    /// </summary>
    /// <param name="propertyName">Name of the property whose time since last update Is required</param>
    /// <returns>Time in seconds since the last sensor update for this property</returns>
    public double TimeSinceLastUpdate(string propertyName)
    {
        // Test for an empty property name, if found, return the time since the most recent update to any sensor
        if (!string.IsNullOrEmpty(propertyName))
        {
            switch (propertyName.Trim().ToLowerInvariant())
            {
                // Return the time for properties that are implemented, otherwise fall through to the MethodNotImplementedException
                case "averageperiod":
                case "cloudcover":
                case "dewpoint":
                case "humidity":
                case "pressure":
                case "rainrate":
                case "skybrightness":
                case "skyquality":
                case "skytemperature":
                case "starfwhm":
                case "temperature":
                case "winddirection":
                case "windgust":
                case "windspeed":
                    // Throw an exception on the properties that are not implemented
                    LogMessage("TimeSinceLastUpdate", $"Property {propertyName} is not implemented");
                    throw new MethodNotImplementedException($"TimeSinceLastUpdate - Property {propertyName} is not implemented");
                default:
                    LogMessage("TimeSinceLastUpdate", $"Invalid sensor name: {propertyName}");
                    throw new InvalidValueException($"TimeSinceLastUpdate - Invalid property name: {propertyName}");
            }
        }

        // Return the time since the most recent update to any sensor
        LogMessage("TimeSinceLastUpdate", $"The time since the most recent sensor update is not implemented");
        throw new MethodNotImplementedException("TimeSinceLastUpdate(" + propertyName + ")");
    }

    /// <summary>
    /// Wind direction at the observatory in degrees
    /// </summary>
    public double WindDirection
    {
        get
        {
            LogMessage("WindDirection", "get - not implemented");
            throw new PropertyNotImplementedException("WindDirection", false);
        }
    }

    /// <summary>
    /// Peak 3 second wind gust at the observatory over the last 2 minutes in m/s
    /// </summary>
    public double WindGust
    {
        get
        {
            LogMessage("WindGust", "get - not implemented");
            throw new PropertyNotImplementedException("WindGust", false);
        }
    }

    /// <summary>
    /// Wind speed at the observatory in m/s
    /// </summary>
    public double WindSpeed
    {
        get
        {
            LogMessage("WindSpeed", "get - not implemented");
            throw new PropertyNotImplementedException("WindSpeed", false);
        }
    }

    #endregion

    #region private methods

    #region calculate the gust strength as the largest wind recorded over the last two minutes

    // save the time and wind speed values
    private Dictionary<DateTime, double> winds = new Dictionary<DateTime, double>();

    private double gustStrength;

    private void UpdateGusts(double speed)
    {
        Dictionary<DateTime, double> newWinds = new Dictionary<DateTime, double>();
        var last = DateTime.Now - TimeSpan.FromMinutes(2);
        winds.Add(DateTime.Now, speed);
        var gust = 0.0;
        foreach (var item in winds)
        {
            if (item.Key > last)
            {
                newWinds.Add(item.Key, item.Value);
                if (item.Value > gust)
                    gust = item.Value;
            }
        }
        gustStrength = gust;
        winds = newWinds;
    }

    #endregion

    #endregion

    //ENDOFINSERTEDFILE
}